lib/block_view.go

```diff
@@ -4826,7 +5378,579 @@ func (bav *UtxoView) _connectUpdateProfile(
+	// considering the transaction metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectUpdateNFT: ")
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorUpdateNFTRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
+		// public key.
+	}
+
+	// Now we are ready to update the NFT. Two things must happen:
+	// 	(1) Update the NFT entry.
+	//  (2) If the NFT entry is being updated to "is not for sale", kill all the bids.
+
+	// Create the updated NFTEntry.
+	newNFTEntry := &NFTEntry{
+		OwnerPKID:         updaterPKID.PKID,
+		NFTPostHash:       txMeta.NFTPostHash,
+		SerialNumber:      txMeta.SerialNumber,
+		IsForSale:         txMeta.IsForSale,
+		MinBidAmountNanos: txMeta.MinBidAmountNanos,
+	}
+	bav._setNFTEntryMappings(newNFTEntry)
+
+	// Iterate over all the NFTBidEntries for this NFT and delete them.
+	bidEntries := bav.GetAllNFTBidEntries(txMeta.NFTPostHash, txMeta.SerialNumber)
+	deletedBidEntries := []*NFTBidEntry{}
+	if bidEntries != nil {
+		for _, bidEntry := range bidEntries {
+			deletedBidEntries = append(deletedBidEntries, bidEntry)
+			bav._deleteNFTBidEntryMappings(bidEntry)
+		}
+	}
+
+	// Add an operation to the list at the end indicating we've connected an NFT bid.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:                 OperationTypeUpdateNFT,
+		PrevNFTEntry:         prevNFTEntry,
+		DeletedNFTBidEntries: deletedBidEntries,
+	})
+
+	return totalInput, totalOutput, utxoOpsForTxn, nil
+}
+
+func (bav *UtxoView) _connectAcceptNFTBid(
+	txn *MsgBitCloutTxn, txHash *BlockHash, blockHeight uint32, verifySignatures bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+
+	// Check that the transaction has the right TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeAcceptNFTBid {
+		return 0, 0, nil, fmt.Errorf("_connectAcceptNFTBid: called with bad TxnType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+	txMeta := txn.TxnMeta.(*AcceptNFTBidMetadata)
+
+	// Verify the NFT entry that is being bid on exists and is on sale.
+	nftKey := MakeNFTKey(txMeta.NFTPostHash, txMeta.SerialNumber)
+	prevNFTEntry := bav.GetNFTEntryForNFTKey(&nftKey)
+	if prevNFTEntry == nil || prevNFTEntry.isDeleted {
+		// We wrap these errors in order to differentiate versus _connectNFTBid().
+		return 0, 0, nil, errors.Wrapf(RuleErrorNFTBidOnNonExistentNFTEntry, "_connectAcceptNFTBid: ")
+	}
+	if !prevNFTEntry.IsForSale {
+		return 0, 0, nil, errors.Wrapf(RuleErrorNFTBidOnNFTThatIsNotForSale, "_connectAcceptNFTBid: ")
+	}
+
+	// Verify that the updater is the owner of the NFT.
+	updaterPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
```

Comment:
just to be safe, we should check that the updaterPKID is not nil and not deleted.
```suggestion
	updaterPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
	if updaterPKID == nil || updaterPKID.isDeleted {
		return 0, 0, nil, errors.Wrapf(err, "_connectAcceptNFTBid: ")
	}
```

Reply:
Same comment as above, not sure what "err" is being wrapped here.

Reply:
yurp - should do a fmt.Errorf thing here.